docs/API_DESIGN.md

```diff
@@ -0,0 +1,652 @@
+      - [Example: Create a `did:dht`, add to the Core Properties \& publish](#example-create-a-diddht-add-to-the-core-properties--publish)
+      - [Example: Instantiate an existing `did:dht`](#example-instantiate-an-existing-diddht)
+      - [Example: Update a `did:dht`](#example-update-a-diddht)
+      - [Example: Resolve a `did:dht`](#example-resolve-a-diddht)
+- [Verifiable Credential's (VCs)](#verifiable-credentials-vcs)
+  - [`NamedIssuer`](#namedissuer)
+  - [`VerifiableCredential`](#verifiablecredential)
+    - [Example: Create a VC \& sign](#example-create-a-vc--sign)
+    - [Example: Verify a VC-JWT](#example-verify-a-vc-jwt)
+- [Presentation Exchange (PEX)](#presentation-exchange-pex)
+  - [`PresentationDefinition`](#presentationdefinition)
+  - [`InputDescriptor`](#inputdescriptor)
+  - [`Constraints`](#constraints)
+  - [`Field`](#field)
+  - [`Optionality`](#optionality)
+  - [`Filter`](#filter)
+
+> [!NOTE]
+> Refer to the [Custom DSL](./CUSTOM_DSL.md) for below syntax definitions.
+
+# Key Material
+
+> [!NOTE]
+> Public & private *key material* are currently strictly represented as [Jwk](#jwk-object-oriented-class), but as the requirement for additional representations (ex: CBOR) present themselves, key material will need to be disintermediated via a polymorphic base class such as `PublicKeyMaterial` (which would expose an instance method for `get_verifier_bytes()`) and `PrivateKeyMaterial` (which would expose instance methods for `to_public_key_material()` and `get_signer_bytes()`), both of which would implement `as_jwk()`, `as_cbor()` and any other concrete representations as instance methods.
+
+## `Jwk`
+
+```pseudocode!
+/// Partial representation of a [JSON Web Key as per RFC7517](https://tools.ietf.org/html/rfc7517).
+/// Note that this is a subset of the spec.
+CLASS Jwk
+  /// Identifies the algorithm intended for use with the key.
+  PUBLIC DATA alg: string
+
+  /// Represents the key type. e.g. EC for elliptic curve, OKP for Edwards curve
+  PUBLIC DATA kty: string
+
+  /// curve name for Elliptic Curve (EC) and Edwards Curve (Ed) keys.
+  /// e.g. secp256k1, Ed25519
+  PUBLIC DATA crv: string
+
+  /// X coordinate for EC keys, or the public key for OKP.
+  PUBLIC DATA x: string
+
+  /// Y coordinate for EC keys.
+  PUBLIC DATA y: string?
+
+  /// Private key component for EC or OKP keys.
+  PUBLIC DATA d: string?
+```
+
+## `InMemoryKeyManager`
+
+```pseudocode!
+/// An encapsulation of key material stored in-memory.
+CLASS InMemoryKeyManager
+  /// Generates new key material and returns the public key represented as a Jwk.
+  METHOD generate_key_material(): Jwk
+
+  /// Returns the Ed25519Signer for the given public key.
+  METHOD get_signer(public_key: Jwk): Ed25519Signer
+
+  /// For importing keys which may be stored somewhere such as environment variables. Return Jwk is the public key for the given private key.
+  METHOD import_key(private_key: Jwk): Jwk
+```
+
+# Digital Signature Algorithm's (DSA)
```

Comment:
I'd suggest a more generic name like `Cryptography` or even just `Crypto` since the code here is already (and will continue to be) broader than _signature algorithms_. Key generation now - and in the future things like encryption or key agreement will make this naming obsolete.

Reply:
The _Custom DSL_ doesn't currently offer prescriptions with respect to namespacing, but I think it may be prudent to consider. I've added namespacing as a missing concept, and I've moved the above section as a sub-section under `Cryptography`, and I've done the same for the VC Data Model 1.1 in your comment below. Thanks!